src/gennylib/include/gennylib/v1/GlobalRateLimiter.hpp

```diff
@@ -0,0 +1,146 @@
+ *
+ * Inspired by
+ * https://github.com/facebook/folly/blob/7c6897aa18e71964e097fc238c93b3efa98b2c61/folly/TokenBucket.h
+ */
+template <typename ClockT = std::chrono::steady_clock>
+class BaseGlobalRateLimiter {
+public:
+    // This should be replaced with std::hardware_destructive_interference_size, which is
+    // part of c++17 but not part of any (major) standard library. Search P0154R1
+    // for more information.
+    //
+    // 64 is the cache line size for recent Intel and AMD processors.
+    static const int CacheLineSize = 64;
+
+    static_assert(ClockT::is_steady, "Clock must be steady");
+    static_assert(std::is_same<typename ClockT::duration, std::chrono::nanoseconds>::value,
+                  "Clock representation must be nano seconds");
+
+public:
+    explicit BaseGlobalRateLimiter(const RateSpec& rs)
+        : _burstSize(rs.operations),
+          _rateNS(rs.per.count()),
+          _lastEmptiedTimeNS{ClockT::now().time_since_epoch().count()} {};
+
+    // No copies or moves.
+    BaseGlobalRateLimiter(const BaseGlobalRateLimiter& other) = delete;
+    BaseGlobalRateLimiter& operator=(const BaseGlobalRateLimiter& other) = delete;
+
+    BaseGlobalRateLimiter(BaseGlobalRateLimiter&& other) = delete;
+    BaseGlobalRateLimiter& operator=(BaseGlobalRateLimiter&& other) = delete;
+
+    ~BaseGlobalRateLimiter() = default;
+
+    /**
+     * Request to consume _burstSize number of tokens from the bucket. Does not block
+     * if the bucket is empty; does block while waiting for concurrent accesses to the
+     * bucket to finish.
+     *
+     * @return bool whether consume() succeeded. The caller is responsible for using an
+     * appropriate back-off strategy if this function returns false.
+     */
+    bool consumeIfWithinRate(const typename ClockT::time_point& now) {
+        // The time the bucket was emptied before this consumeIfWithinRate() call.
+        int64_t curEmptiedTime = _lastEmptiedTimeNS.load();
+
+        // The time the bucket was emptied after this consumeIfWithinRate() call.
+        int64_t newEmptiedTime;
+
+        auto nowInTicks = now.time_since_epoch().count();
+        do {
+            newEmptiedTime = curEmptiedTime + _rateNS;
+
+            // If the new emptied time is in the future, the bucket is empty. Return early.
+            if (nowInTicks < newEmptiedTime) {
+                return false;
+            }
+            // Use the "weak" version for performance at the expense of false negatives (i.e.
+            // `compare_exchange` not comparing equal when it should).
+        } while (!_lastEmptiedTimeNS.compare_exchange_weak(curEmptiedTime, newEmptiedTime));
+
+        return true;
+    }
+
+    constexpr int64_t getBurstSize() const {
+        return _burstSize;
+    }
+
+    constexpr int64_t getRate() const {
+        return _rateNS;
+    }
+
+    constexpr int64_t getNumUsers() const {
+        return _numUsers;
+    }
+
+    void addUser() {
+        _numUsers++;
+    }
```

Comment:
since the rate-limiter itself doesn't ever use this value, can it instead be stored in the thing that knows the value (the workload context)? workload context could instead have `map<string,pair<limiter,count>>` and the `getRateLimiter` could increment that count?

Either that or please add comment about how this value is intended to be used.